### PR TITLE
added select flagged primitive

### PR DESCRIPTION
**Pull request by makart19**
added select flagged primitive (in classic daal it was named as partition).

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/select/select_dpc.cpp [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    const ndview<Data, 1>& in,
+    ndview<Data, 1>& out,
+    std::int64_t& selected_elem_count,
+    const event_vector& deps) {
+    ONEDAL_ASSERT(out.has_mutable_data());
+    ONEDAL_ASSERT(in.get_count() == out.get_count());
+
+    if (in.get_count() > de::limits<integer_t>::max()) {
+        /// CHange exception name
+        throw domain_error(dal::detail::error_messages::invalid_number_of_elements_to_process());
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/select/select_dpc.cpp:228]: Is it TODO?

File: cpp/oneapi/dal/backend/primitives/select/test/perf_select_dpc.cpp [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        Flag* mask_ptr = detail::host_allocator<Flag>().allocate(val.get_count());
+        for (Integer el = 0; el < elem_count; el++) {
+            val_ptr[el] = distr(rng);
+            mask_ptr[el] = val_ptr[el] < pivot ? 1 : 0;
+        }
+        val.assign(this->get_queue(), val_ptr, val.get_count()).wait_and_throw();
+        mask.assign(this->get_queue(), mask_ptr, val.get_count()).wait_and_throw();
+
+        detail::host_allocator<Float>().deallocate(val_ptr, val.get_count());
+        detail::host_allocator<Flag>().deallocate(mask_ptr, val.get_count());
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/select/test/perf_select_dpc.cpp:111]: It sounds like ndarray need to be created on host and next copied to device. All you need is `to_device`.
```c++
auto generate_uniform(std::int64_t count, Float a, Float b, int seed) const {
    std::mt19937 rng(seed);
    std::uniform_real_distribution<Float> distr(a, b);

    const auto val = ndarray<Float, 1>::empty(count);
    const auto mask = ndarray<Flag, 1>::empty(count); 

    Float* val_ptr = val.get_mutable_data();
    Flag* mask_ptr = mask.get_mutable_data();

    for (std::int64_t el = 0; el < count; el++) {
        val_ptr[el] = distr(rng);
        mask_ptr[el] = val_ptr[el] < pivot ? 1 : 0;
    } 
    
    auto q = this->get_queue();
    return std::make_tuple(val.to_device(q), mask.to_device(q));
}
```

Comment (rlnx) [reply cpp/oneapi/dal/backend/primitives/select/test/perf_select_dpc.cpp:111]: In general, I see a lot of code copy-pasted in perf and functional tests. I think we could have common header, where base fuxture for these tests is defined. It's not problem of this particular primitive, but also sort and reduction to some extent. 

We shouldn't address this right now, but we need to keep in mind and plan refactoring iteration.

File: cpp/oneapi/dal/backend/primitives/select/select.hpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    using base_t = select_flagged_base<Data, Flag>;
+
+public:
+    using mask_t = typename base_t::mask_t;
+
+    /// Performs initialization of auxiliary variables and required auxiliary buffers
+    /// @param[in]  queue The queue
+    select_flagged(const sycl::queue& queue) : select_flagged_base<Data, Flag>(queue){};
+    select_flagged(const select_flagged&) = delete;
+    ~select_flagged(){};
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/select/select.hpp:149]: Can be omitted

File: cpp/oneapi/dal/backend/primitives/select/select.hpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    using base_t = select_flagged_base<Data, Flag>;
+
+public:
+    using mask_t = typename base_t::mask_t;
+
+    /// Performs initialization of auxiliary variables and required auxiliary buffers
+    /// @param[in]  queue The queue
+    select_flagged_index(const sycl::queue& queue) : select_flagged_base<Data, Flag>(queue){};
+    select_flagged_index(const select_flagged_index&) = delete;
+    ~select_flagged_index(){};
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/select/select.hpp:187]: Can be omitted

File: cpp/oneapi/dal/backend/primitives/select/select.hpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    static_assert(std::numeric_limits<Data>::is_integer);
+
+    using base_t = select_flagged_base<Data, Flag>;
+
+public:
+    using mask_t = typename base_t::mask_t;
+
+    /// Performs initialization of auxiliary variables and required auxiliary buffers
+    /// @param[in]  queue The queue
+    select_flagged_index(const sycl::queue& queue) : select_flagged_base<Data, Flag>(queue){};
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/select/select.hpp:185]: Extra `;`
```suggestion
    select_flagged_index(const sycl::queue& queue) : select_flagged_base<Data, Flag>(queue){}
```

File: cpp/oneapi/dal/backend/primitives/select/select.hpp [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    static_assert(std::is_floating_point<Data>::value);
+
+    using base_t = select_flagged_base<Data, Flag>;
+
+public:
+    using mask_t = typename base_t::mask_t;
+
+    /// Performs initialization of auxiliary variables and required auxiliary buffers
+    /// @param[in]  queue The queue
+    select_flagged(const sycl::queue& queue) : select_flagged_base<Data, Flag>(queue){};
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/select/select.hpp:147]: Extra `;`
```suggestion
    select_flagged(const sycl::queue& queue) : select_flagged_base<Data, Flag>(queue){}
```

File: cpp/oneapi/dal/backend/primitives/select/select_dpc.cpp (hunk 1 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/select/select_dpc.cpp]: no, already changed, forget to remove comment.

File: cpp/oneapi/dal/backend/primitives/select/select.hpp (hunk 3 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/select/select.hpp]: removed

File: cpp/oneapi/dal/backend/primitives/select/select.hpp (hunk 4 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/select/select.hpp]: removed

File: cpp/oneapi/dal/backend/primitives/select/select.hpp (hunk 5 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/select/select.hpp]: fixed

File: cpp/oneapi/dal/backend/primitives/select/select.hpp (hunk 6 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/select/select.hpp]: fixed

File: cpp/oneapi/dal/backend/primitives/select/test/perf_select_dpc.cpp (hunk 2 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/select/test/perf_select_dpc.cpp]: added method to device, rewrote this code snippet accordingly, but keep "fill_uniform" naming for the case when generation can be done as separate step, for example in perf tests, where it is used for filling each element of vector of ndarrays.

File: cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        auto policy = de::data_parallel_policy{ this->get_queue() };
+        Data* val_ptr = de::data_parallel_allocator<Data>{ policy }.allocate(elem_count_);
+        Flag* mask_ptr = detail::data_parallel_allocator<Flag>{ policy }.allocate(elem_count_);
+
+        auto val = ndarray<Data, 1>::wrap(val_ptr,
+                                          { dummy_elem_count },
+                                          detail::make_default_delete<Data>(policy));
+        auto mask = ndarray<Flag, 1>::wrap(mask_ptr,
+                                           { dummy_elem_count },
+                                           detail::make_default_delete<Flag>(policy));
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp:88]: Use `ndarray::empty(q, { dummy_elem_count }, sycl::usm::alloc::device)`

File: cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    void check_select(ndarray<Data, 1>& in, ndarray<Flag, 1>& mask) {
+        auto policy = de::data_parallel_policy{ this->get_queue() };
+        Data* out_ptr = de::data_parallel_allocator<Data>{ policy }.allocate(elem_count_);
+
+        auto out = ndarray<Data, 1>::wrap(out_ptr,
+                                          { in.get_count() },
+                                          detail::make_default_delete<Data>(policy));
+        std::int64_t total_sum_res;
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp:100]: Use `ndarray::empty`

File: cpp/oneapi/dal/backend/primitives/selection/test/perf_select_flagged_dpc.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+        for (Integer el = 0; el < elem_count; el++) {
+            val_ptr[el] = distr(rng);
+            mask_ptr[el] = val_ptr[el] < pivot ? 1 : 0;
+        }
+
+        auto& q = this->get_queue();
+
+        val = val_host.to_device(q);
+        mask = mask_host.to_device(q);
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/selection/test/perf_select_flagged_dpc.cpp:94]: It looks like, there is no need for separation on allocation and initialization any more. You can do the following:
```c++
TEST(...) {
    auto [val, mask] = this->init_uniform(elem_count, -25., 25.);
}
``` 
instead
```c++
TEST(...) {
    auto [val, mask] = this->allocate_arrays(elem_count);
    this->fill_uniform(val, mask, -25., 25.);
}
```

Comment (rlnx) [reply cpp/oneapi/dal/backend/primitives/selection/test/perf_select_flagged_dpc.cpp:94]: The same for `init_vector_of_arrays` memory allocation and initialization can be done in the same loop.

File: cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp (hunk 7 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp:88]: This block for creating dummy ndarray for testing overflow checks, therefore  i can't use ndarray::empty here.

File: cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp (hunk 8 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/selection/test/select_flagged_overflow_dpc.cpp:100]: This block for creating dummy ndarray for testing overflow checks, therefore  i can't use ndarray::empty here.

File: cpp/oneapi/dal/backend/primitives/selection/test/perf_select_flagged_dpc.cpp (hunk 9 above)
Comment (makart19) [reply cpp/oneapi/dal/backend/primitives/selection/test/perf_select_flagged_dpc.cpp]: replaced to_device with val.assign(q, val_host);

**Review by rlnx — APPROVED** [review]

**Pull request merged**